File: assets/scripts/simple_python_example.py
# TwoDEngine Simple Python Example
# This script demonstrates data exchange between C++ and Python using files

def read_data_from_cpp(filename):
    """
    Read data from a file written by C++

    Args:
        filename: The name of the file to read

    Returns:
        A list of dictionaries containing the data points
    """
    data_points = []

    try:
        with open(filename, 'r') as file:
            # Read the number of data points
            num_points = int(file.readline().strip())

            # Read each data point
            for _ in range(num_points):
                line = file.readline().strip()
                name, value, x, y = line.split(',')
                data_points.append({
                    'name': name,
                    'value': int(value),
                    'x': float(x),
                    'y': float(y)
                })

        print(f"Read {len(data_points)} data points from C++")

    except Exception as e:
        print(f"Error reading data from file: {e}")

    return data_points

def write_data_to_cpp(data_points, filename):
    """
    Write data to a file for C++ to read

    Args:
        data_points: A list of dictionaries containing the data points
        filename: The name of the file to write
    """
    try:
        with open(filename, 'w') as file:
            # Write the number of data points
            file.write(f"{len(data_points)}\n")

            # Write each data point
            for point in data_points:
                file.write(f"{point['name']},{point['value']},{point['x']},{point['y']}\n")

        print(f"Wrote {len(data_points)} data points for C++")

    except Exception as e:
        print(f"Error writing data to file: {e}")

def process_data(data_points):
    """
    Process the data points

    Args:
        data_points: A list of dictionaries containing the data points

    Returns:
        The processed data points
    """
    print("Processing data in Python...")

    # Print the data points
    for point in data_points:
        print(f"Python sees: {point['name']}, value={point['value']}, coordinates=({point['x']}, {point['y']})")

    # Modify the data points
    for point in data_points:
        point['value'] *= 2  # Double the value
        point['x'] += 0.5    # Add 0.5 to x
        point['y'] += 0.5    # Add 0.5 to y

    # Add a new data point
    data_points.append({
        'name': 'Point C',
        'value': 30,
        'x': 5.0,
        'y': 6.0
    })

    # Calculate statistics
    total = sum(point['value'] for point in data_points)
    average = total / len(data_points) if data_points else 0

    print(f"Sum calculated in Python: {total}")
    print(f"Average calculated in Python: {average}")

    return data_points

def main():
    print("Simple Python Example - Python Side")
    print("===================================")

    # Read data from C++
    data_points = read_data_from_cpp("cpp_data.txt")

    # Process the data
    processed_data = process_data(data_points)

    # Write data back to C++
    write_data_to_cpp(processed_data, "python_data.txt")

    print("Python processing completed successfully!")

if __name__ == "__main__":
    main()

File: examples/SimplePythonExample.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <cstdlib>

// A simple data structure to demonstrate C++ to Python data exchange
struct DataPoint {
    std::string name;
    int value;
    double x;
    double y;

    DataPoint(const std::string& n, int v, double x_val, double y_val)
        : name(n), value(v), x(x_val), y(y_val) {}

    std::string toString() const {
        return "DataPoint: " + name + ", value=" + std::to_string(value) +
               ", coordinates=(" + std::to_string(x) + ", " + std::to_string(y) + ")";
    }
};

// A simple data processor class
class DataProcessor {
private:
    std::vector<DataPoint> dataPoints;

public:
    void addDataPoint(const DataPoint& point) {
        dataPoints.push_back(point);
    }

    void clearData() {
        dataPoints.clear();
    }

    const std::vector<DataPoint>& getData() const {
        return dataPoints;
    }

    int calculateSum() const {
        int sum = 0;
        for (const auto& point : dataPoints) {
            sum += point.value;
        }
        return sum;
    }

    double calculateAverage() const {
        if (dataPoints.empty()) return 0.0;
        return static_cast<double>(calculateSum()) / dataPoints.size();
    }
};

// Function to write data to a file for Python to read
void writeDataToFile(const DataProcessor& processor, const std::string& filename) {
    std::ofstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Failed to open file for writing: " << filename << std::endl;
        return;
    }

    // Write the number of data points
    file << processor.getData().size() << std::endl;

    // Write each data point
    for (const auto& point : processor.getData()) {
        file << point.name << "," << point.value << "," << point.x << "," << point.y << std::endl;
    }

    file.close();
}

// Function to read data from a file that Python wrote
void readDataFromFile(DataProcessor& processor, const std::string& filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Failed to open file for reading: " << filename << std::endl;
        return;
    }

    processor.clearData();

    // Read the number of data points
    int numPoints;
    file >> numPoints;
    file.ignore(); // Skip the newline

    // Read each data point
    for (int i = 0; i < numPoints; ++i) {
        std::string line;
        std::getline(file, line);

        std::stringstream ss(line);
        std::string name;
        int value;
        double x, y;

        std::getline(ss, name, ',');
        ss >> value;
        ss.ignore(); // Skip the comma
        ss >> x;
        ss.ignore(); // Skip the comma
        ss >> y;

        processor.addDataPoint(DataPoint(name, value, x, y));
    }

    file.close();
}

int main() {
    std::cout << "Simple Python Example" << std::endl;
    std::cout << "====================" << std::endl;

    // Create C++ data
    DataProcessor processor;
    processor.addDataPoint(DataPoint("Point A", 10, 1.0, 2.0));
    processor.addDataPoint(DataPoint("Point B", 20, 3.0, 4.0));

    // Print initial data
    std::cout << "Initial data in C++:" << std::endl;
    for (const auto& point : processor.getData()) {
        std::cout << point.toString() << std::endl;
    }
    std::cout << "Sum: " << processor.calculateSum() << std::endl;
    std::cout << "Average: " << processor.calculateAverage() << std::endl;
    std::cout << std::endl;

    // Write data to file for Python to read
    std::cout << "Writing data to file for Python..." << std::endl;
    writeDataToFile(processor, "cpp_data.txt");

    // Run Python script to process the data
    std::cout << "Running Python script to process data..." << std::endl;
    int result = system("..\\python-3.10.11\\python.exe assets/scripts/simple_python_example.py");
    if (result != 0) {
        std::cerr << "Error running Python script. Error code: " << result << std::endl;
        return 1;
    }

    // Read processed data from Python
    std::cout << "Reading processed data from Python..." << std::endl;
    readDataFromFile(processor, "python_data.txt");

    // Print processed data
    std::cout << "Data after Python processing:" << std::endl;
    for (const auto& point : processor.getData()) {
        std::cout << point.toString() << std::endl;
    }
    std::cout << "Sum: " << processor.calculateSum() << std::endl;
    std::cout << "Average: " << processor.calculateAverage() << std::endl;

    std::cout << "\nSimple Python example completed successfully!" << std::endl;
    return 0;
}
